File: packages/typia/src/factories/LiteralFactory.ts
import ts from "typescript";

import { ExpressionFactory } from "./ExpressionFactory";
import { IdentifierFactory } from "./IdentifierFactory";

export namespace LiteralFactory {
  export const write = (input: any): ts.Expression => {
    if (input === null) return ts.factory.createNull();
    else if (ts.isArrowFunction(input)) return input;
    else if (ts.isCallExpression(input)) return input;
    else if (ts.isIdentifier(input)) return input;
    else if (input instanceof Array) return writeArray(input);
    else if (typeof input === "object") return writeObject(input);
    else if (typeof input === "boolean") return writeBoolean(input);
    else if (typeof input === "bigint") return writeBigint(input);
    else if (typeof input === "number") return writeNumber(input);
    else if (typeof input === "string") return writeStrinng(input);
    // unreachable code
    else if (typeof input === "function")
      return ts.factory.createIdentifier("undefined");
    else
      throw new TypeError("Error on LiteralFactory.generate(): unknown type.");
  };

  const writeObject = (obj: object): ts.ObjectLiteralExpression =>
    ts.factory.createObjectLiteralExpression(
      Object.entries(obj)
        .filter((tuple) => tuple[1] !== undefined)
        .map(([key, value]) =>
          ts.factory.createPropertyAssignment(
            IdentifierFactory.identifier(key),
            write(value),
          ),
        ),
      true,
    );

  const writeArray = (array: any[]): ts.ArrayLiteralExpression =>
    ts.factory.createArrayLiteralExpression(array.map(write), true);

  const writeBoolean = (value: boolean): ts.Expression =>
    value ? ts.factory.createTrue() : ts.factory.createFalse();

  const writeNumber = (value: number): ts.Expression =>
    ExpressionFactory.number(value);

  const writeBigint = (value: bigint): ts.Expression =>
    ExpressionFactory.bigint(value);

  const writeStrinng = (value: string): ts.StringLiteral =>
    ts.factory.createStringLiteral(value);
}

File: packages/core/src/factories/LiteralFactory.go
package factories

import "fmt"

type literalFactory struct{}

var LiteralFactory literalFactory

func (literalFactory) Write(input any) (Expression, error) {
	switch value := input.(type) {
	case nil:
		return Null(), nil
	case Expression:
		return value, nil
	case []any:
		items := make([]Expression, 0, len(value))
		for _, elem := range value {
			written, err := LiteralFactory.Write(elem)
			if err != nil {
				return Expression{}, err
			}
			items = append(items, written)
		}
		return Node("arrayLiteral", Field("elements", items)), nil
	case map[string]any:
		properties := make([]ObjectProperty, 0, len(value))
		for key, elem := range value {
			if elem == nil {
				continue
			}
			written, err := LiteralFactory.Write(elem)
			if err != nil {
				return Expression{}, err
			}
			properties = append(properties, ObjectProperty{Key: IdentifierFactory.Identifier(key), Value: written})
		}
		return Node("objectLiteral", Field("properties", properties)), nil
	case bool:
		if value {
			return True(), nil
		}
		return False(), nil
	case int:
		return ExpressionFactory.Number(float64(value)), nil
	case int64:
		return ExpressionFactory.Number(float64(value)), nil
	case float64:
		return ExpressionFactory.Number(value), nil
	case string:
		return StringLiteral(value), nil
	default:
		return Expression{}, fmt.Errorf("Error on LiteralFactory.generate(): unknown type %T", input)
	}
}
